docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,368 @@
+
+#### HelmChartProxy CRD
+
+The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 
+
+##### Spec
+
+- `clusterSelector`: A key/value pair used as a label selector. Clusters with a matching label will be included in orchestration, i.e. installing the Helm chart on them and keeping them up to date.
+- `repoURL`: The URL of the Helm chart repository.
+- `chartName`: The name of the Helm chart.
+- `releaseName`: The name of the Helm release.
+- `version`: The version of the Helm chart.
+- `values`: A map of key/value pairs specifying values to be passed to the Helm chart. Go templating is supported to reference fields from each workload cluster spec.
+
+##### Status
+
+- `matchingClusters`: A list of object references to workload clusters matching the cluster selector.
+- `ready`: A boolean that is true when the list of HelmReleaseProxies are up to date.
+- `failureReason`: A string pointer indicating the error message, if applicable.
+
+```yaml
+apiVersion: addons.cluster.x-k8s.io/v1beta1
+kind: HelmChartProxy
+metadata:
+  name: cloud-provider-azure-helmchartproxy
+spec:
+  clusterSelector:
+    key: cloudProviderAzureChart
+    value: enabled
+  repoURL: https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo
+  chartName: cloud-provider-azure
+  version: 1.23.8
+  releaseName: cloud-provider-azure
+  values:
+    infra.clusterName: "{{ .Cluster.Name }}"
+    cloudControllerManager.containerResourceManagement.limitsCPU: "{{ add .ControlPlane.Spec.Replicas 10 }}"
+status:
+  matchingClusters:
+  - apiVersion: cluster.x-k8s.io/v1beta1
+    kind: Cluster
+    name: default-15210
+    namespace: default
+  ready: true
+```
+
+#### HelmReleaseProxy CRD
+
+The second CRD, HelmReleaseProxy, is responsible for installing and orchestrating the Helm release on a selected workload Cluster. The HelmReleaseProxy is created as part of the parent HelmChartProxy flow. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).
+
+##### Spec
+
+- `clusterRef`: An object reference to the Cluster the Helm release is installed on.
+- `repoURL`: The URL of the Helm chart repository.
+- `chartName`: The name of the Helm chart.
+- `releaseName`: The name of the Helm release.
+- `version`: The version of the Helm chart.
+- `values`: A map of key/value pairs specifying values to be used when installing the Helm release. Note that this map is populated with the resolved values from the Cluster.
+
+##### Status
+
+- `ready`: A boolean that indicates when the Helm release is successfully installed and up to date with the HelmReleaseProxy spec.
+- `failureReason`: A string pointer indicating the error message, if applicable.
```

Comment:
let's consider adding Conditions to better surface granular status details. FailureReason is technically deprecated and should theoretically be removed from other CAPI CRDs eventually (https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md)

Reply:
Yep, adding conditions is on my TODO list, though I didn't realize FailureReason is deprecated. Is FailureMessage in the same position?

Reply:
Yes